README.md
# mongoose-field-encryption

[![Build Status](https://travis-ci.org/wheresvic/mongoose-field-encryption.svg?branch=master)](https://travis-ci.org/wheresvic/mongoose-field-encryption) [![Coverage Status](https://coveralls.io/repos/github/wheresvic/mongoose-field-encryption/badge.svg?branch=master)](https://coveralls.io/github/wheresvic/mongoose-field-encryption?branch=master)

A simple symmetric encryption plugin for individual fields. The goal of this plugin is to encrypt data but still allow searching over fields with string values. This plugin relies on the Node `crypto` module. Encryption and decryption happen transparently during save and find.

As of the stable 1.0.0 release, this plugin works on individual fields of any type. However, note that for non-string fields, the original value is set to undefined after encryption. This is because if the schema has defined a field as an array, it would not be possible to replace it with a string value.

Also consider [mongoose-encryption](https://github.com/joegoldbeck/mongoose-encryption) if you have other requirements.

## How it works

Encryption is performed using `AES-256-CBC`. To encrypt, the relevant fields are encrypted with the provided secret + random salt. The generated salt and the resulting encrypted value is concatenated together using a `:` character and the final string is put in place of the actual value for `string` values. An extra `boolean` field with the prefix `__enc_` is added to the document which indicates if the provided field is encrypted or not.

Fields which are either objects or of a different type are converted to strings using `JSON.stringify` and the value stored in an extra marker field of type `string` with a naming scheme of `__enc_` as prefix and `_d` as suffix on the original field name. The original field is then set to `undefined`. Please note that this might break any custom validation and application of this plugin on non-string fields needs to be done with care.

## Requirements

- Node `>=4.4.7`
- MongoDB `>=2.6.10`
- Mongoose `>=4.0.0`

## Installation

`npm install mongoose-field-encryption`

## Usage

Keep your secret a secret. Ideally it should only live as an environment variable but definitely not stored anywhere in your repository.

### Basic

For example, given a schema as follows:

```js
const mongoose                = require('mongoose');
const mongooseFieldEncryption = require('mongoose-field-encryption').fieldEncryption;
const Schema                  = mongoose.Schema;

const Post = new Schema({
  title: String,
  message: String,
  references: {
    author: String,
    date: Date
  }
});

Post.plugin(mongooseFieldEncyption, {fields: ['message', 'references'], secret: 'some secret key'});
```

The resulting documents will have the following format:

```js
{
  _id: ObjectId,
  title: String,
  message: String, // encrypted salt and hex value as string, e.g. 9d6a0ca4ac2c80fc84df0a06de36b548:cee57185fed78c055ed31ca6a8be9bf20d303283200a280d0f4fc8a92902e0c1
  __enc_message: true, // boolean marking if the field is encrypted or not
  references: undefined, // encrypted object set to undefined
  __enc_references: true, // boolean marking if the field is encrypted or not
  __enc_references_d: String // encrypted salt and hex object value as string, e.g. 6df2171f25fd1d32adc4a4059f867a82:5909152856cf9cdb7dc32c6af321c8fe69390c359c6b19d967eaa6e7a0a97216
}
```

`find` works transparently and you can make new documents as normal, but you should not use the `lean` option on a find if you want the fields of the document to be decrypted. `findOne`, `findById` and `save` also all work as normal. `update` works _only for string fields_ and you would also need to manually set the `__enc_` field value to false if you're updating an encrypted field.

From the mongoose package documentation: _Note that findAndUpdate/Remove do not execute any hooks or validation before making the change in the database. If you need hooks and validation, first query for the document and then save it._

Note that as of `1.2.0` release, support for `findOneAndUpdate` has also been added. Note that you would need to specifically set the encryption field marker for it to be encrypted. For example:

```js
Post.findOneAndUpdate(
  { _id: postId },
  { $set: { message: "snoop", __enc_message: false } }
);
```

The above also works for non-string fields. See changelog for more details.

Also note that if you manually set the value `__enc_` prefix field to true then the encryption is not run on the corresponding field and this may result in the plain value being stored in the db.

### Options

- `fields` (required): an array list of the required fields
- `secret` (required): a string cipher which is used to encrypt the data (don't lose this!)
- `useAes256Ctr` (optional, default `false`): a boolean indicating whether the older `aes-256-ctr` algorithm should be used. Note that this is strictly a backwards compatibility feature and for new installations it is recommended to leave this at default.

### Static methods

For performance reasons, once the document has been encrypted, it remains so. The following methods are thus added to the schema:

- `encryptFieldsSync()`: synchronous call that encrypts all fields as given by the plugin options
- `decryptFieldsSync()`: synchronous call that decrypts encrypted fields as given by the plugin options
- `stripEncryptionFieldMarkers()`: synchronous call that removes the encryption field markers (useful for returning documents without letting the user know that something was encrypted)

Multiple calls to the above methods have no effect, i.e. once a field is encrypted and the `__enc_` marker field value is set to true then the ecrypt operation is ignored. Same for the decrypt operation. Of course if the field markers have been removed via the `stripEncryptionFieldMarkers()` call, then the encryption will be executed if invoked.

### Searching

To enable searching over the encrypted fields the `encrypt` and `decrypt` methods have also been exposed.

```js
const fieldEncryption = require('mongoose-field-encryption')
const encrypted = fieldEncryption.encrypt('some text', 'secret'));
const decrypted = fieldEncryption.decrypt(encrypted, 'secret')); // decrypted = 'some text'
```

## Testing

1. Install dependencies with `npm install` and [install mongo](http://docs.mongodb.org/manual/installation/) if you don't have it yet.
2. Start mongo with `mongod`.
3. Run tests with `npm test`. Additionally you can pass your own mongodb uri as an environment variable if you would like to test against your own database, for e.g. `URI='mongodb://username:password@127.0.0.1:27017/mongoose-field-encryption-test' npm test`

## Publishing

- `npm version patch,minor,major`
- `npm publish`

## Changelog

### 2.0.0

- Use `cipheriv` instead of plain `cipher`, [#17](https://github.com/wheresvic/mongoose-field-encryption/issues/17).
  
  Note that this might break any _fixed_ search capability as the encrypted values are now based on a random salt.
  
  Also note that while this version maintains backward compatibility, i.e. decryption will automatically fall back to using the `aes-256-ctr` algorithm, any further updates will lead to the value being encrypted with the salt. In order to fully maintain backwards compatibilty, an new option `useAes256Ctr` has been introduced (default `false`), which can be set to `true` to continue using the plugin as before. It is highly recommended to start using the newer algorithm however, see issue for more details.

### 1.2.0

- Added support for `findOneAndUpdate` [https://github.com/wheresvic/mongoose-field-encryption/pull/20](https://github.com/wheresvic/mongoose-field-encryption/pull/20)

### 1.1.0

- Added support for mongoose 5 [https://github.com/wheresvic/mongoose-field-encryption/pull/16](https://github.com/wheresvic/mongoose-field-encryption/pull/16).
- Removed mongoose dependency, moved to `peerDependencies`.
- Formatted source code using prettier.


CHANGELOG.md
# CHANGELOG

See README for changelog


test/test-db.js
"use strict";

const expect = require("chai").expect;

const Promise = require("bluebird");
const mongoose = require("mongoose");
mongoose.Promise = Promise;
mongoose.set("bufferCommands", false);

const fieldEncryptionPlugin = require("../lib/mongoose-field-encryption").fieldEncryption;

const uri = process.env.URI || "mongodb://127.0.0.1:27017/mongoose-field-encryption-test";

describe("mongoose-field-encryption plugin db", function() {
  this.timeout(5000);

  let NestedFieldEncryptionSchema = new mongoose.Schema({
    toEncryptString: { type: String, required: true },
    toEncryptObject: {
      nested: String
    },
    toEncryptArray: [],
    toEncryptDate: Date
  });

  NestedFieldEncryptionSchema.plugin(fieldEncryptionPlugin, {
    fields: ["toEncryptString", "toEncryptObject", "toEncryptArray", "toEncryptDate"],
    secret: "icanhazcheezburger" // should ideally be process.env.SECRET
  });

  let NestedFieldEncryption = mongoose.model("NestedFieldEncryption", NestedFieldEncryptionSchema);

  before(() => {
    return mongoose.connect(
      uri,
      { useNewUrlParser: true, promiseLibrary: Promise, autoIndex: false }
    );
  });

  after(() => {
    mongoose.connection.close();
  });

  function getSut() {
    let sut = new NestedFieldEncryption({
      toEncryptString: "hide me!",
      toEncryptObject: {
        nested: "some stuff to encrypt"
      },
      toEncryptArray: [1, 2, 3],
      toEncryptDate: new Date(1485641048338)
    });

    return sut;
  }

  function expectEncryptionValues(sut) {
    expect(sut.__enc_toEncryptString).to.be.true;

    expect(sut.toObject().toEncryptObject).to.be.undefined;
    expect(sut.__enc_toEncryptObject).to.be.true;

    expect(sut.toObject().toEncryptArray).to.be.undefined;
    expect(sut.__enc_toEncryptArray).to.be.true;

    expect(sut.toObject().toEncryptDate).to.be.undefined;
    expect(sut.__enc_toEncryptDate).to.be.true;
  }

  function expectDecryptionValues(found) {
    expect(found.toEncryptString).to.equal("hide me!");
    expect(found.__enc_toEncryptString).to.be.false;

    expect(JSON.stringify(found.toEncryptObject)).to.equal('{"nested":"some stuff to encrypt"}');
    expect(found.__enc_toEncryptObject).to.be.false;
    expect(found.__enc_toEncryptObject_d).to.equal("");

    expect(JSON.stringify(found.toEncryptArray)).to.equal("[1,2,3]");
    expect(found.__enc_toEncryptArray).to.be.false;
    expect(found.__enc_toEncryptArray_d).to.equal("");

    expect(JSON.stringify(found.toEncryptDate)).to.equal('"2017-01-28T22:04:08.338Z"');
    expect(found.__enc_toEncryptDate).to.be.false;
    expect(found.__enc_toEncryptDate_d).to.equal("");
  }

  it("should encrypt fields on save and decrypt fields on findById", () => {
    // given
    let sut = getSut();

    // when
    return sut
      .save()
      .then(() => {
        // then
        expectEncryptionValues(sut);

        return NestedFieldEncryption.findById(sut._id);
      })
      .then(found => {
        expectDecryptionValues(found);
      });
  });

  it("should encrypt fields on save and decrypt fields on findOne", () => {
    // given
    let sut = getSut();

    // when
    return sut
      .save()
      .then(() => {
        expectEncryptionValues(sut);
        return NestedFieldEncryption.findOne({ _id: sut._id });
      })
      .then(found => {
        expectDecryptionValues(found);
      });
  });

  it("should store encrypted fields as plaintext on findOneAndUpdate", () => {
    // given
    let sut = getSut();

    // when
    return sut
      .save()
      .then(() => {
        expectEncryptionValues(sut);

        return NestedFieldEncryption.findOneAndUpdate(
          { _id: sut._id },
          {
            $set: { toEncryptString: "snoop", __enc_toEncryptString: false }
          },
          { new: true }
        );
      })
      .then(found => {
        // then
        expect(found.__enc_toEncryptString).to.be.false;
        expect(found.toEncryptString).to.equal("snoop");
      });
  });

  it("should encrypt string fields on update", () => {
    // given
    let sut = getSut();

    // when
    return sut
      .save()
      .then(() => {
        expectEncryptionValues(sut);

        return NestedFieldEncryption.update(
          { _id: sut._id },
          { $set: { toEncryptString: "snoop", __enc_toEncryptString: false } }
        );
      })
      .then(() => {
        return NestedFieldEncryption.findById(sut._id);
      })
      .then(found => {
        // then
        expect(found.__enc_toEncryptString).to.be.false;
        expect(found.toEncryptString).to.equal("snoop");
      });
  });

  it("should encrypt string fields on update without $set", () => {
    // given
    let sut = getSut();

    // when
    return sut
      .save()
      .then(() => {
        expectEncryptionValues(sut);

        return NestedFieldEncryption.update(
          { _id: sut._id },
          { toEncryptString: "snoop" }
        );
      })
      .then(() => {
        return NestedFieldEncryption.findById(sut._id);
      })
      .then(found => {
        // then
        expect(found.__enc_toEncryptString).to.be.false;
        expect(found.toEncryptString).to.equal("snoop");
      });
  });

  it("should encrypt string fields on fineOneAndUpdate", () => {
    // given
    let sut = getSut();

    // when
    return sut
      .save()
      .then(() => {
        expectEncryptionValues(sut);

        return NestedFieldEncryption.findOneAndUpdate(
          { _id: sut._id },
          { $set: { toEncryptString: "snoop", __enc_toEncryptString: false } }
        );
      })
      .then(() => {
        return NestedFieldEncryption.findById(sut._id);
      })
      .then(found => {
        // then
        expect(found.__enc_toEncryptString).to.be.false;
        expect(found.toEncryptString).to.equal("snoop");
      });
  })

  it("should encrypt non string fields on update", () => {
    // given
    let sut = getSut();

    // when
    return sut
      .save()
      .then(() => {
        expectEncryptionValues(sut);

        return NestedFieldEncryption.update(
          {
            _id: sut._id
          },
          {
            $set: {
              toEncryptObject: { nested: "snoop" },
              __enc_toEncryptObject: false
            }
          }
        );
      })
      .then(() => {
        return NestedFieldEncryption.findById(sut._id);
      })
      .then(found => {
        expect(found.toEncryptObject.nested).to.eql("snoop");
      });
  });

  it("should encrypt non string fields on fineOneAndUpdate without $set", () => {
    // given
    let sut = getSut();

    // when
    return sut
      .save()
      .then(() => {
        expectEncryptionValues(sut);

        return NestedFieldEncryption.findOneAndUpdate(
          {
            _id: sut._id
          },
          {
            toEncryptObject: { nested: "snoop" }
          }
        );
      })
      .then(() => {
        return NestedFieldEncryption.findById(sut._id);
      })
      .then(found => {
        expect(found.toEncryptObject.nested).to.eql("snoop");
      });
  });

  it ("should not encrypt already encrypted fields", () => {
    // given
    let sut = getSut();

    // when
    return sut
      .save()
      .then(() => {
        expectEncryptionValues(sut);

        return NestedFieldEncryption.update(
          {
            _id: sut._id
          },
          {
            $set: {
              toEncryptString: "already encrypted string",
              __enc_toEncryptObject: true
            }
          }
        );
      })
      .then(() => {
        return NestedFieldEncryption.findById(sut._id);
      })
      .then(found => {
        expect(found.toEncryptString).to.eql("already encrypted string");
      });
  })
});


test/test-setup.js
"use strict";

const expect = require("chai").expect;

const Promise = require("bluebird");
const mongoose = require("mongoose");
mongoose.Promise = Promise;

const fieldEncryptionPlugin = require("../lib/mongoose-field-encryption").fieldEncryption;

describe("mongoose-field-encryption plugin setup", () => {
  let FieldEncryptionSchema = new mongoose.Schema({
    a: { type: String, required: true }
  });

  it("should not initialize plugin without a secret", done => {
    try {
      // when
      FieldEncryptionSchema.plugin(fieldEncryptionPlugin);
    } catch (err) {
      // then
      expect(err.message).to.equal("missing required secret");
      return done();
    }

    expect.fail("Should not have initialized plugin");
    done();
  });

  it("should initialize plugin without any fields", done => {
    // when
    FieldEncryptionSchema.plugin(fieldEncryptionPlugin, {
      secret: "icanhazcheezburger"
    });

    // then
    done();
  });
});


lib/mongoose-field-encryption.js
"use strict";

const crypto = require("crypto");

const algorithm = "aes-256-cbc";
const deprecatedAlgorithm = "aes-256-ctr";
const encryptedFieldNamePrefix = "__enc_";
const encryptedFieldDataSuffix = "_d";

const encryptAes256Ctr = function(text, secret) {
  const cipher = crypto.createCipher(deprecatedAlgorithm, secret);
  let crypted = cipher.update(text, "utf8", "hex");
  crypted += cipher.final("hex");
  return crypted;
};

const decryptAes256Ctr = function(encryptedHex, secret) {
  const decipher = crypto.createDecipher(deprecatedAlgorithm, secret);
  let dec = decipher.update(encryptedHex, "hex", "utf8");
  dec += decipher.final("utf8");
  return dec;
};

const encrypt = function(clearText, secret) {
  const iv = crypto.randomBytes(16);
  const cipher = crypto.createCipheriv(algorithm, secret, iv);
  const encrypted = cipher.update(clearText);
  const finalBuffer = Buffer.concat([encrypted, cipher.final()]);
  const encryptedHex = iv.toString("hex") + ":" + finalBuffer.toString("hex");
  return encryptedHex;
};

/**
 * Decryption has a default fallback for the deprecated algorithm
 * 
 * @param {*} encryptedHex
 * @param {*} secret
 */
const decrypt = function(encryptedHex, secret) {
  const encryptedArray = encryptedHex.split(":");

  // maintain backwards compatibility
  if (encryptedArray.length === 1) {
    return decryptAes256Ctr(encryptedArray[0], secret);
  }

  const iv = new Buffer.from(encryptedArray[0], "hex");
  const encrypted = new Buffer.from(encryptedArray[1], "hex");
  const decipher = crypto.createDecipheriv(algorithm, secret, iv);
  const decrypted = decipher.update(encrypted);
  const clearText = Buffer.concat([decrypted, decipher.final()]).toString();
  return clearText;
};

const fieldEncryption = function(schema, options) {
  if (!options || !options.secret) {
    throw new Error("missing required secret");
  }

  const useAes256Ctr = options.useAes256Ctr || false;
  const fieldsToEncrypt = options.fields || [];

  const hash = crypto.createHash("sha256");
  hash.update(options.secret);

  const secret = useAes256Ctr ? options.secret : hash.digest("hex").substring(0, 32);
  const encryptionStrategy = useAes256Ctr ? encryptAes256Ctr : encrypt;

  // add marker fields to schema
  for (let field of fieldsToEncrypt) {
    const encryptedFieldName = encryptedFieldNamePrefix + field;
    const encryptedFieldData = encryptedFieldName + encryptedFieldDataSuffix;
    const schemaField = {};

    schemaField[encryptedFieldName] = { type: Boolean };
    schemaField[encryptedFieldData] = { type: String };
    schema.add(schemaField);
  }

  //
  // local methods
  //

  // for mongoose 4/5 compatibility
  const defaultNext = function defaultNext(err) {
    if (err) {
      throw err;
    }
  };

  function getCompatitibleNextFunc(next) {
    if (typeof next !== "function") {
      return defaultNext;
    }
    return next;
  }

  function getCompatibleData(next, data) {
    // in mongoose5, 'data' field is undefined
    if (!data) {
      return next;
    }
    return data;
  }

  function encryptFields(obj, fields, secret) {
    for (let field of fields) {
      const encryptedFieldName = encryptedFieldNamePrefix + field;
      const encryptedFieldData = encryptedFieldName + encryptedFieldDataSuffix;
      const fieldValue = obj[field];

      if (!obj[encryptedFieldName] && fieldValue) {
        if (typeof fieldValue === "string") {
          // handle strings separately to maintain searchability
          const value = encryptionStrategy(fieldValue, secret);
          obj[field] = value;
        } else {
          const value = encryptionStrategy(JSON.stringify(fieldValue), secret);
          obj[field] = undefined;
          obj[encryptedFieldData] = value;
        }

        obj[encryptedFieldName] = true;
      }
    }
  }

  function decryptFields(obj, fields, secret) {
    for (let field of fields) {
      const encryptedFieldName = encryptedFieldNamePrefix + field;
      const encryptedFieldData = encryptedFieldName + encryptedFieldDataSuffix;

      if (obj[encryptedFieldName]) {
        if (obj[encryptedFieldData]) {
          const encryptedValue = obj[encryptedFieldData];

          obj[field] = JSON.parse(decrypt(encryptedValue, secret));
          obj[encryptedFieldName] = false;
          obj[encryptedFieldData] = "";
        } else {
          // handle strings separately to maintain searchability
          const encryptedValue = obj[field];

          obj[field] = decrypt(encryptedValue, secret);
          obj[encryptedFieldName] = false;
        }
      }
    }
  }

  function updateHook(_next) {
    const next = getCompatitibleNextFunc(_next);
    for (let field of fieldsToEncrypt) {
      const encryptedFieldName = encryptedFieldNamePrefix + field;
      this._update.$set = this._update.$set || {};
      const plainTextValue = this._update.$set[field] || this._update[field];
      const encryptedFieldValue = this._update.$set[encryptedFieldName] || this._update[encryptedFieldName];

      if (!encryptedFieldValue && plainTextValue) {
        let updateObj = {};
        if (typeof plainTextValue === "string" || plainTextValue instanceof String) {
          const encryptedData = encryptionStrategy(plainTextValue, secret);

          updateObj[field] = encryptedData;
          updateObj[encryptedFieldName] = true;
        } else {
          const encryptedFieldData = encryptedFieldName + encryptedFieldDataSuffix;

          updateObj[field] = undefined;
          updateObj[encryptedFieldData] = encryptionStrategy(JSON.stringify(plainTextValue), secret);
          updateObj[encryptedFieldName] = true;
        }
        this.update({}, Object.keys(this._update.$set).length > 0 ? { $set: updateObj } : updateObj);
      }
    }

    next();
  }

  //
  // static methods
  //

  schema.methods.stripEncryptionFieldMarkers = function() {
    for (let field of fieldsToEncrypt) {
      let encryptedFieldName = encryptedFieldNamePrefix + field;
      let encryptedFieldData = encryptedFieldName + encryptedFieldDataSuffix;

      this.set(encryptedFieldName, undefined);
      this.set(encryptedFieldData, undefined);
    }
  };

  schema.methods.decryptFieldsSync = function() {
    decryptFields(this, fieldsToEncrypt, secret);
  };

  schema.methods.encryptFieldsSync = function() {
    encryptFields(this, fieldsToEncrypt, secret);
  };

  //
  // hooks
  //

  schema.pre("init", function(_next, _data) {
    const next = getCompatitibleNextFunc(_next);
    const data = getCompatibleData(_next, _data);
    try {
      decryptFields(data, fieldsToEncrypt, secret);
      next();
    } catch (err) {
      next(err);
    }
  });

  schema.pre("save", function(_next) {
    const next = getCompatitibleNextFunc(_next);

    try {
      encryptFields(this, fieldsToEncrypt, secret);
      next();
    } catch (err) {
      next(err);
    }
  });

  schema.pre("findOneAndUpdate", updateHook);

  schema.pre("update", updateHook);
};

module.exports.fieldEncryption = fieldEncryption;
module.exports.encrypt = encrypt;
module.exports.decrypt = decrypt;
module.exports.encryptAes256Ctr = encryptAes256Ctr;
